prboom2/src/SDL/i_sound.c

```diff
@@ -163,21 +164,74 @@ static void stopchan(int i)
 //
 static int addsfx(int sfxid, int channel, const unsigned char *data, size_t len)
 {
+  channel_info_t *ci = channelinfo + channel;
+
   stopchan(channel);
 
-  channelinfo[channel].data = data;
-  /* Set pointer to end of raw data. */
-  channelinfo[channel].enddata = channelinfo[channel].data + len - 1;
-  channelinfo[channel].samplerate = (channelinfo[channel].data[3] << 8) + channelinfo[channel].data[2];
-  channelinfo[channel].data += 8; /* Skip header */
+  if (len > 44 && memcmp(data, "RIFF", 4) == 0 && memcmp(data + 8, "WAVEfmt ", 8) == 0)
+  {
+    SDL_RWops *RWops;
+    SDL_AudioSpec wav_spec;
+    Uint8 *wav_buffer = NULL;
+    int bits, samplelen;
+
+    RWops = SDL_RWFromMem(data, len);
+
+    if (SDL_LoadWAV_RW(RWops, 1, &wav_spec, &wav_buffer, &samplelen) == NULL)
+    {
+      lprintf(LO_WARN, "Could not open wav file: %s\n", SDL_GetError());
+      return channel;
+    }
+
+    if (wav_spec.channels != 1)
+    {
+      lprintf(LO_WARN, "Only mono WAV file is supported");
+      SDL_FreeWAV(wav_buffer);
+      return channel;
+    }
+
+    if (!SDL_AUDIO_ISINT(wav_spec.format))
+    {
+      lprintf(LO_WARN, "WAV file in unsupported format");
+      SDL_FreeWAV(wav_buffer);
+      return channel;
+    }
+
+    bits = SDL_AUDIO_BITSIZE(wav_spec.format);
+    if (bits != 8 && bits != 16)
+    {
+      lprintf(LO_WARN, "Only 8 or 16 bit WAV files are supported");
+      SDL_FreeWAV(wav_buffer);
+      return channel;
+    }
+
+    Z_Free(data);
```

Comment:
I'm even more confused now 😅
This method is receiving a piece of data that is a lump. The lump itself is locked and allocated in zone memory. Every time the sound effect is played, that lump data will be sent to this method. Why are we freeing the lump data here?

Reply:
Because if that lump is a WAV file, it might be compressed and needs to be decoded.

Reply:
My question is why are we freeing the raw lump data in this function. It's an unexpected side effect to me (outside this method, the data is locked, but then it gets freed before it returns).

Reply:
I understand you are allocating a new buffer, but outside this method we have a pointer to data that now doesn't exist, which is not a safe practice 😅

What I'm telling you is that all sound effect data in dsda-doom is cached when the program launches and does not need to be freed manually at any time. We use lump locking / unlocking to manage lump data (which takes care of cleanup if and when it is needed).

Reply:
You can think of the locking as a way of managing zone memory and swapping the tags based on the number of locks. This is done so one place in the code base can lock a lump, and another can, and they will unlock it when they are done. Then no part needs to know if any other part has locked the data, as the zone is managing the lock counts.

Reply:
If for some reason another part of the code base has locked the sound effect lump, and you manually free it, then that part now has an invalid pointer but doesn't know it.

Reply:
> What I'm telling you is that all sound effect data in dsda-doom is cached when the program launches and does not need to be freed manually at any time. We use lump locking / unlocking to manage lump data (which takes care of cleanup if and when it is needed).

But in the case at hand the sound *lump* does not contain the actual sound *data*, but some WAV format stuff that first needs to be converted into your actual sound data format. And once this conversion is done, you will never need the sound *lump* again.

Reply:
Right, sorry, Woof does it differently. So is it safe just write `ci->data = wav_buffer`? Or we need to to call `SDL_FreeWAV(wav_buffer)` somewhere?

Reply:
> But in the case at hand the sound _lump_ does not contain the actual sound _data_, but some WAV format stuff that first needs to be converted into your actual sound data format. And once this conversion is done, you will never need the sound _lump_ again.

There are two problems here.

This data has been locked by another part of the program. Whether or not that is "correct" or meaningful is besides the point. Freeing data that is reserved somewhere else is a fundamentally dangerous action and relies on assumptions that may or may not be incorrect either now or in the future.

Unless I'm mistaken, `addsfx` is called by `I_StartSound` every time a sound is played, which means the lump data is accessed every time the sound plays and this whole conversion is happening every time a sound is played. Is there somewhere that short circuits this flow of events if it has happened before?

Reply:
In Woof, all sounds are pre-cached at start-up unless sound pitch randomization is enabled. Maybe I am mixing up its concepts with PrBoom+/DSDA. Sorry if I added even more confusion.

Reply:
> Unless I'm mistaken, `addsfx` is called by `I_StartSound` every time a sound is played, which means the lump data is accessed every time the sound plays and this whole conversion is happening every time a sound is played. Is there somewhere that short circuits this flow of events if it has happened before?

I think we need some sort of caching data structure. I'll try to add it a little later.

Reply:
👍